expected-consensus.md

```diff
@@ -39,22 +40,65 @@ Expected Consensus is a probabilistic Byzantine fault-tolerant consensus protoco
 level, it operates by running a leader election every round in which, on expectation, one 
 participant may be eligible to submit a block. All valid blocks submitted in a given round form a `TipSet`. Every block in a TipSet adds weight to its chain. The 'best' chain is the one with the highest weight, which is to say that the fork choice rule is to choose the heaviest known chain. For more details on this, see [Chain Selection](#chain-selection).
 
-The basic algorithm is as follows:
+The basic algorithms are as follows:
 
-- Collect all incoming blocks.
-- Check their validity as shown in [the mining spec.](./mining.md#block-validation)
-- For each block that is valid, and not known to be based on a bad chain, place it into a `ChainTipsManager.`
-- To check if a miner has won for round `N`:
-  - Sample the secure chain randomness at round `N-K` to receive a ticket (see [Tickets](#tickets) for details). `K` is called the randomness lookback parameter, currently set to 1.
-  - Compute an `ElectionProof` using this ticket as input.
-  - If the hash of the `ElectionProof` is less than the ratio of the miner's power to the total power in the network at the end of round `N-L`, the miner has a winning ticket. `L` is called the committee lookback parameter, currently set to `1`.
-  - The `ElectionProof` will be included in the new block header.
-- If the miner is a winner, generate a new block:
-  - Select the smallest ticket from the heaviest TipSet at round `N-1`.
-  - Generate a new ticket from it for inclusion in the new block.
-  - Generate a new block as shown in [the mining spec](./mining.md#block-creation).
+For each block received over the network, `OnBlockReceived` is called. `IsValid` 
+
+```go
+func OnBlockReceived(blk Block) {
+    // The exact definition of IsValid depends on the protocol
+    // For Filecoin, see mining.md
+    if IsValid(blk) {
+        ChainTipsMgr.Add(blk)
+    }
+    
+    // Received an invalid block!
+}
+```
+
+Separately, another process is running `Mine` to attempt to generate blocks.
+```go
+func Mine(minerKey PrivateKey) {
+    for r := range rounds { // for each round
+        bestTipset, tickets := ChainTipsMgr.GetBestTipsetAtRound(r-1)
+        
+        ticket := GenerateTicket(minerKey, bestTipset)
+        tickets.Append(ticket)
+        
+        // Generate an election proof and check if we win
+        // Note: even if we don't win, we want to make a ticket
+        // in case we need to mine a null round
+        win, proof := CheckIfWinnerAtRound(minerKey, r, bestTipset)
+        if win {
+            GenerateAndBroadcastBlock(bestTipset, tickets, proof)
+        } else {
+            // Even if we don't win, add our ticket to the tracker in 
+            // case we need to mine on it later.
+            ChainTipsMgr.AddFailedTicket(bestTipset, tickets)
```

Comment:
nit: this may be clearer if you actually clear the tickets array on a win (ie line 74), and otherwise have it be initialized outside the for loop, which is to say the miner always maintains a running ticket array. Whenever we can include these in a block (ie on a win) or a new block comes in, it gets reset, otherwise it grows.

Nvm i've convinced myself that capturing all of that logic in Mine is bad. Better to do it through the chainTipsMgr.

But in that case I may nonetheless have it clear tickets after the win, (just as you would taking in a block)

Reply:
I think the logic of when to clear stored 'losing' tickets should be left up to the chaintipsmgr, it has to do other maintenance of that set, so doing some here some there feels odd

Reply:
Fair enough, though it's one of the main questions that comes to my mind when reading the spec. Maybe worth mentioning this explicitly (ie that logic is handled by the CTSMgr) or giving a quick written intuition here?